File: packages/sdk-py/src/grantex/revocations/_feed.py
"""The revocation feed client (PRD G-6).

It takes a snapshot of everything currently revoked, then follows the live
stream in a background thread, so ``enforce()`` can deny a call on a grant that
was revoked seconds ago without a network round trip per call.

The important part is what happens when it cannot keep up. The feed records
when it last heard from the auth service; if that is longer ago than
``stale_after``, it reports itself stale and ``enforce()`` denies rather than
allowing calls on information it knows may be out of date.
"""

from __future__ import annotations

import json
import threading
import time
from dataclasses import dataclass
from typing import Any, Literal, Optional

import httpx

from ._set import RevocationEntry, RevocationMatch, RevokedSet

FeedUnavailableReason = Literal["disabled", "not_ready", "unauthorized", "network"]

DEFAULT_STALE_AFTER = 5.0
DEFAULT_RECONNECT_DELAY = 0.5
_MAX_RECONNECT_DELAY = 30.0
# The stream may sit idle between revocations, but heartbeats arrive often;
# connecting and sending must not hang.
_STREAM_TIMEOUT = httpx.Timeout(connect=10.0, read=None, write=10.0, pool=10.0)
_REQUEST_TIMEOUT = httpx.Timeout(10.0)


@dataclass(frozen=True)
class RevocationFeedState:
    """What the feed knows, and whether it can be trusted."""

    synced: bool
    fresh_at: float
    cursor: int
    known: int
    unavailable: Optional[FeedUnavailableReason]


class RevocationFeed:
    """Follows the auth service's revocation feed in a background thread."""

    def __init__(
        self,
        base_url: str,
        api_key: str,
        *,
        stale_after: float = DEFAULT_STALE_AFTER,
        reconnect_delay: float = DEFAULT_RECONNECT_DELAY,
        transport: Literal["stream", "poll"] = "stream",
    ) -> None:
        self._base_url = base_url.rstrip("/")
        self._headers = {"Authorization": f"Bearer {api_key.strip()}", "Accept": "application/json"}
        self.stale_after = stale_after
        self._reconnect_delay = reconnect_delay
        self._transport = transport
        self._set = RevokedSet()
        self._lock = threading.Lock()
        self._changed = threading.Condition(self._lock)
        self._stop = threading.Event()
        self._thread: Optional[threading.Thread] = None
        self._synced = False
        self._fresh_at = 0.0
        self._cursor = 0
        self._unavailable: Optional[FeedUnavailableReason] = None
        self._response: Optional[httpx.Response] = None

    # ── state ────────────────────────────────────────────────────────────────

    def state(self) -> RevocationFeedState:
        with self._lock:
            return RevocationFeedState(
                synced=self._synced,
                fresh_at=self._fresh_at,
                cursor=self._cursor,
                known=self._set.size,
                unavailable=self._unavailable,
            )

    def is_fresh(self, now: Optional[float] = None) -> bool:
        """Synced recently enough to be trusted."""
        moment = now if now is not None else time.time()
        with self._lock:
            return (
                self._synced
                and self._unavailable is None
                and moment - self._fresh_at <= self.stale_after
            )

    def match(
        self,
        *,
        grant_id: Optional[str] = None,
        token_id: Optional[str] = None,
        parent_grant_id: Optional[str] = None,
    ) -> Optional[RevocationMatch]:
        """Why this credential must not be used, according to what the feed knows."""
        return self._set.match(
            grant_id=grant_id, token_id=token_id, parent_grant_id=parent_grant_id
        )

    # ── lifecycle ────────────────────────────────────────────────────────────

    def start(self) -> None:
        """Start following the feed. Idempotent."""
        with self._lock:
            if self._thread is not None and self._thread.is_alive():
                return
            self._stop.clear()
            self._thread = threading.Thread(target=self._run, name="grantex-revocation-feed", daemon=True)
            self._thread.start()

    def stop(self) -> None:
        """Stop following, and forget what was known."""
        self._stop.set()
        with self._lock:
            response = self._response
            thread = self._thread
            self._thread = None
            self._synced = False
            self._fresh_at = 0.0
            self._changed.notify_all()
        if response is not None:
            try:
                response.close()
            except Exception:  # noqa: BLE001 - closing a dead stream is not an error
                pass
        if thread is not None:
            thread.join(timeout=5)
        self._set.clear()

    def ready(self, timeout: Optional[float] = None) -> bool:
        """Wait until the feed is fresh, or the timeout passes.

        Returns whether it is fresh: ``False`` means callers must fail closed.
        """
        self.start()
        deadline = time.time() + (timeout if timeout is not None else self.stale_after)
        while True:
            if self.is_fresh():
                return True
            with self._lock:
                if self._unavailable is not None:
                    return False
                remaining = deadline - time.time()
                if remaining <= 0:
                    break
                self._changed.wait(min(remaining, 0.05))
        return self.is_fresh()

    # ── the loop ─────────────────────────────────────────────────────────────

    def _touch(self) -> None:
        with self._lock:
            self._fresh_at = time.time()
            self._unavailable = None
            self._changed.notify_all()

    def _fail(self, reason: FeedUnavailableReason) -> None:
        with self._lock:
            self._unavailable = reason
            self._synced = False
            self._changed.notify_all()

    def _run(self) -> None:
        delay = self._reconnect_delay
        with httpx.Client(headers=self._headers, timeout=_REQUEST_TIMEOUT) as client:
            while not self._stop.is_set():
                try:
                    self._snapshot(client)
                    delay = self._reconnect_delay
                    if self._transport == "stream":
                        self._stream(client)
                    else:
                        self._poll(client)
                except Exception as exc:  # noqa: BLE001 - every failure is a reason to fail closed
                    self._fail(_classify(exc))
                    self._stop.wait(delay)
                    delay = min(delay * 2, _MAX_RECONNECT_DELAY)

    def _snapshot(self, client: httpx.Client) -> None:
        page_token: Optional[str] = None
        cursor = 0
        while True:
            params: dict[str, Any] = {}
            if page_token:
                params["pageToken"] = page_token
            response = client.get(f"{self._base_url}/v1/revocations", params=params)
            response.raise_for_status()
            page = response.json()
            self._set.apply_all(RevocationEntry.from_dict(item) for item in page.get("entries", []))
            cursor = int(page.get("cursor", 0) or 0)
            page_token = page.get("nextPageToken")
            if not page_token:
                break
        self._set.prune()
        with self._lock:
            self._cursor = cursor
            self._synced = True
        self._touch()

    def _stream(self, client: httpx.Client) -> None:
        with self._lock:
            cursor = self._cursor
        with client.stream(
            "GET",
            f"{self._base_url}/v1/revocations/stream",
            params={"since": cursor},
            timeout=_STREAM_TIMEOUT,
        ) as response:
            response.raise_for_status()
            with self._lock:
                self._response = response
            try:
                event = "message"
                data = ""
                for line in response.iter_lines():
                    if self._stop.is_set():
                        return
                    if line == "":
                        self._handle_event(event, data)
                        event, data = "message", ""
                        continue
                    if line.startswith("event:"):
                        event = line[6:].strip()
                    elif line.startswith("data:"):
                        data += line[5:].strip()
            finally:
                with self._lock:
                    self._response = None

    def _handle_event(self, event: str, data: str) -> None:
        if event in ("heartbeat", "ready"):
            self._touch()
            return
        if event != "revocation" or not data:
            return
        try:
            entry = RevocationEntry.from_dict(json.loads(data))
        except (ValueError, TypeError):
            # A line we cannot read is not evidence that nothing is revoked;
            # the next heartbeat decides freshness.
            return
        self._set.apply(entry)
        with self._lock:
            if entry.seq > self._cursor:
                self._cursor = entry.seq
        self._touch()

    def _poll(self, client: httpx.Client) -> None:
        wait = max(1, int(self.stale_after / 2))
        while not self._stop.is_set():
            with self._lock:
                cursor = self._cursor
            response = client.get(
                f"{self._base_url}/v1/revocations",
                params={"since": cursor, "wait": wait},
                timeout=httpx.Timeout(connect=10.0, read=wait + 10.0, write=10.0, pool=10.0),
            )
            response.raise_for_status()
            page = response.json()
            entries = [RevocationEntry.from_dict(item) for item in page.get("entries", [])]
            self._set.apply_all(entries)
            with self._lock:
                self._cursor = max(
                    [int(page.get("cursor", cursor) or cursor), *[entry.seq for entry in entries], cursor]
                )
            self._touch()


def _classify(exc: Exception) -> FeedUnavailableReason:
    status = getattr(getattr(exc, "response", None), "status_code", None)
    if status == 404:
        return "disabled"
    if status == 503:
        return "not_ready"
    if status in (401, 403):
        return "unauthorized"
    return "network"

File: packages/sdk-ts/src/revocations/feed.ts
/**
 * The revocation feed client (PRD G-6).
 *
 * It takes a snapshot of everything currently revoked, then follows the live
 * stream, so `enforce()` can deny a call on a grant that was revoked seconds
 * ago without a network round trip per call.
 *
 * The important part is what happens when it cannot keep up. The feed records
 * when it last heard from the auth service; if that is longer ago than
 * `staleAfterMs`, it reports itself stale and `enforce()` denies rather than
 * allowing calls on information it knows may be out of date.
 */
import type { HttpClient } from '../http.js';
import { RevokedSet, type CredentialRef, type RevocationEntry, type RevocationMatch } from './set.js';

export interface RevocationFeedOptions {
  /**
   * How long the feed may go without hearing from the auth service before it
   * is stale and calls fail closed. Default 5000 ms.
   */
  staleAfterMs?: number;
  /** Delay before reconnecting a dropped stream. Default 500 ms. */
  reconnectDelayMs?: number;
  /** `stream` (Server-Sent Events) or `poll` (long polling). Default `stream`, falling back to `poll`. */
  transport?: 'stream' | 'poll';
}

export type FeedUnavailableReason = 'disabled' | 'not_ready' | 'unauthorized' | 'network';

export interface RevocationFeedState {
  /** A snapshot has been read at least once. */
  synced: boolean;
  /**
   * When the feed last heard from the auth service, on a monotonic clock
   * (0 if never). Comparable with `performance.now()`, not with `Date.now()`.
   */
  freshAt: number;
  /** Feed position of the last entry applied. */
  cursor: number;
  /** Identifiers currently known to be revoked or suspended. */
  known: number;
  /** Why the feed is not usable, if it is not. */
  unavailable: FeedUnavailableReason | null;
}

interface FeedPage {
  entries: RevocationEntry[];
  cursor: number;
  nextPageToken?: string | null;
  snapshot?: boolean;
}

/**
 * Freshness is measured on a monotonic clock: a wall-clock jump backwards
 * must not make a stale feed look current.
 */
const monotonicNow = (): number =>
  (typeof performance !== 'undefined' && typeof performance.now === 'function' ? performance.now() : Date.now());

const DEFAULT_STALE_AFTER_MS = 5_000;
const DEFAULT_RECONNECT_DELAY_MS = 500;
const MAX_RECONNECT_DELAY_MS = 30_000;
const PRUNE_INTERVAL_MS = 60_000;

export class RevocationFeed {
  readonly #http: HttpClient;
  readonly #set = new RevokedSet();
  readonly #staleAfterMs: number;
  readonly #reconnectDelayMs: number;
  readonly #transport: 'stream' | 'poll';

  #running = false;
  #synced = false;
  #freshAt = 0;
  #cursor = 0;
  #unavailable: FeedUnavailableReason | null = null;
  #controller: AbortController | null = null;
  #loop: Promise<void> | null = null;
  #reader: ReadableStreamDefaultReader<Uint8Array> | null = null;
  #prunedAt = 0;
  #waiters: Array<() => void> = [];

  constructor(http: HttpClient, options: RevocationFeedOptions = {}) {
    this.#http = http;
    this.#staleAfterMs = options.staleAfterMs ?? DEFAULT_STALE_AFTER_MS;
    this.#reconnectDelayMs = options.reconnectDelayMs ?? DEFAULT_RECONNECT_DELAY_MS;
    this.#transport = options.transport ?? 'stream';
  }

  get staleAfterMs(): number {
    return this.#staleAfterMs;
  }

  state(): RevocationFeedState {
    return {
      synced: this.#synced,
      freshAt: this.#freshAt,
      cursor: this.#cursor,
      known: this.#set.size,
      unavailable: this.#unavailable,
    };
  }

  /** Synced recently enough to be trusted. */
  isFresh(now: number = monotonicNow()): boolean {
    return this.#synced && this.#unavailable === null && now - this.#freshAt <= this.#staleAfterMs;
  }

  /** Why this credential must not be used, according to what the feed knows. */
  match(ref: CredentialRef, now: number = Date.now()): RevocationMatch | null {
    return this.#set.match(ref, now);
  }

  /** Start following the feed. Idempotent. */
  start(): void {
    if (this.#running) return;
    this.#running = true;
    this.#controller = new AbortController();
    this.#loop = this.#run().catch(() => { /* the loop reports through state() */ });
  }

  /** Stop following, and forget what was known (a restarted feed re-reads the snapshot). */
  async stop(): Promise<void> {
    this.#running = false;
    this.#controller?.abort();
    // An open stream is not interrupted by aborting a fetch that already
    // resolved, so release the reader too.
    await this.#reader?.cancel().catch(() => { /* already closed */ });
    const loop = this.#loop;
    this.#loop = null;
    this.#synced = false;
    this.#freshAt = 0;
    this.#set.clear();
    this.#wake();
    if (loop) await loop;
  }

  /**
   * Wait until the feed is fresh, or the timeout passes. Returns whether it is
   * fresh: `false` means callers must fail closed.
   */
  async ready(timeoutMs: number = this.#staleAfterMs): Promise<boolean> {
    this.start();
    const deadline = monotonicNow() + timeoutMs;
    for (;;) {
      if (this.isFresh()) return true;
      if (this.#unavailable !== null || monotonicNow() >= deadline) return this.isFresh();
      await new Promise<void>((resolve) => {
        const timer = setTimeout(resolve, Math.min(50, Math.max(1, deadline - monotonicNow())));
        timer.unref?.();
        this.#waiters.push(() => {
          clearTimeout(timer);
          resolve();
        });
      });
    }
  }

  #wake(): void {
    const waiters = this.#waiters;
    this.#waiters = [];
    for (const waiter of waiters) waiter();
  }

  #touch(): void {
    this.#freshAt = monotonicNow();
    // Entries whose credential has expired can no longer be used; dropping
    // them here bounds the set without a timer of its own.
    if (this.#freshAt - this.#prunedAt > PRUNE_INTERVAL_MS) {
      this.#set.prune();
      this.#prunedAt = this.#freshAt;
    }
    this.#unavailable = null;
    this.#wake();
  }

  async #run(): Promise<void> {
    let delay = this.#reconnectDelayMs;
    while (this.#running) {
      try {
        await this.#snapshot();
        delay = this.#reconnectDelayMs;
        if (this.#transport === 'stream') {
          const streamed = await this.#stream();
          if (!streamed) await this.#poll();
          // A stream that ended cleanly (a server restart, a proxy timeout)
          // must not be reconnected in a tight loop.
          else if (this.#running) await sleep(this.#reconnectDelayMs, this.#controller?.signal);
        } else {
          await this.#poll();
        }
      } catch (err) {
        this.#unavailable = classify(err);
        this.#synced = false;
        this.#wake();
        await sleep(delay, this.#controller?.signal);
        delay = Math.min(delay * 2, MAX_RECONNECT_DELAY_MS);
      }
    }
  }

  async #snapshot(): Promise<void> {
    let pageToken: string | null | undefined;
    let cursor = 0;
    // Collected first and swapped in once every page has arrived. A snapshot
    // is the whole truth about what is revoked *now*, so applying it on top
    // of what the set already held kept anything resumed while this client
    // was disconnected — denied until it expired. Swapping only after the
    // last page means a failure part way through leaves the previous set
    // intact rather than a half-built one.
    const entries: RevocationEntry[] = [];
    do {
      // `pageToken` is empty only on the first pass; the loop condition ends
      // it otherwise. The previous null/undefined pair was half redundant,
      // which is what the static analysis was pointing at.
      const query = pageToken ? `?pageToken=${encodeURIComponent(pageToken)}` : '';
      const page = await this.#http.get<FeedPage>(`/v1/revocations${query}`);
      entries.push(...page.entries);
      cursor = page.cursor;
      pageToken = page.nextPageToken;
    } while (pageToken);
    this.#set.replaceAll(entries);
    this.#set.prune();
    this.#cursor = cursor;
    this.#synced = true;
    this.#touch();
  }

  /** Follow the Server-Sent Events stream. Returns false if streaming is not usable here. */
  async #stream(): Promise<boolean> {
    const response = await this.#http.rawGet(
      `/v1/revocations/stream?since=${this.#cursor}`,
      this.#controller?.signal,
    );
    if (!response.ok) throw httpError(response.status);
    if (!response.body) return false;

    const reader = response.body.getReader();
    this.#reader = reader;
    const decoder = new TextDecoder();
    let buffer = '';
    // A half-open socket delivers nothing and never ends. Heartbeats are due
    // every second, so silence for twice the staleness bound means the
    // connection is gone: drop it and let the loop reconnect.
    const silenceLimit = Math.max(2 * this.#staleAfterMs, 1_000);
    let watchdog = setTimeout(() => void reader.cancel().catch(() => { /* already closed */ }), silenceLimit);
    watchdog.unref?.();
    try {
      while (this.#running) {
        const { done, value } = await reader.read();
        if (done) break;
        clearTimeout(watchdog);
        watchdog = setTimeout(() => void reader.cancel().catch(() => { /* already closed */ }), silenceLimit);
        watchdog.unref?.();
        buffer += decoder.decode(value, { stream: true });
        let index = buffer.indexOf('\n\n');
        while (index !== -1) {
          this.#handleEvent(buffer.slice(0, index));
          buffer = buffer.slice(index + 2);
          index = buffer.indexOf('\n\n');
        }
      }
    } finally {
      clearTimeout(watchdog);
      this.#reader = null;
      reader.releaseLock();
      await response.body.cancel().catch(() => { /* already closed */ });
    }
    return true;
  }

  #handleEvent(block: string): void {
    let event = 'message';
    let data = '';
    for (const line of block.split('\n')) {
      if (line.startsWith('event:')) event = line.slice(6).trim();
      else if (line.startsWith('data:')) data += line.slice(5).trim();
    }
    if (event === 'heartbeat' || event === 'ready') {
      this.#touch();
      return;
    }
    if (event !== 'revocation' || data.length === 0) return;
    try {
      const entry = JSON.parse(data) as RevocationEntry;
      this.#set.apply(entry);
      if (entry.seq > this.#cursor) this.#cursor = entry.seq;
      this.#touch();
    } catch {
      // A line we cannot read is not evidence that nothing is revoked; the
      // next heartbeat decides freshness.
    }
  }

  /** Long-poll fallback for runtimes without streaming response bodies. */
  async #poll(): Promise<void> {
    const wait = Math.max(1, Math.floor(this.#staleAfterMs / 2 / 1000));
    while (this.#running) {
      const page = await this.#http.get<FeedPage>(`/v1/revocations?since=${this.#cursor}&wait=${wait}`);
      this.#set.applyAll(page.entries);
      if (page.cursor > this.#cursor) this.#cursor = page.cursor;
      for (const entry of page.entries) if (entry.seq > this.#cursor) this.#cursor = entry.seq;
      this.#touch();
    }
  }
}

function httpError(status: number): Error {
  const error = new Error(`revocation feed responded ${status}`);
  (error as Error & { status?: number }).status = status;
  return error;
}

function classify(err: unknown): FeedUnavailableReason {
  const status = (err as { status?: number; statusCode?: number } | null)?.statusCode
    ?? (err as { status?: number } | null)?.status;
  if (status === 404) return 'disabled';
  if (status === 503) return 'not_ready';
  if (status === 401 || status === 403) return 'unauthorized';
  return 'network';
}

async function sleep(ms: number, signal?: AbortSignal): Promise<void> {
  if (signal?.aborted) return;
  await new Promise<void>((resolve) => {
    const timer = setTimeout(resolve, ms);
    timer.unref?.();
    signal?.addEventListener('abort', () => {
      clearTimeout(timer);
      resolve();
    }, { once: true });
  });
}
